api/classes/publisher.js

```diff
@@ -98,53 +105,133 @@ function modifyFiles(action, project, user) {
+  }).fetch()
+  .then(function(publishedProject) {
+    if (!publishedProject) {
+      return;
+    }
+    return publishedProject;
+  })
+  .catch(reject);
 }
 
-exports.publish = function(project) {
-  return getUser(project)
-    .then(function(user) {
-      return modifyFiles(uploadFile, project, user);
-    })
-    .then(function(url) {
-      return project.set({
-        publish_url: url
-      }).save();
-    })
-    .then(function() {
-      log.info(
-        'Successfully published ' +
-        project.get('title') +
-        ' to ' + project.get('publish_url')
-      );
-    })
-    .catch(function(e) {
-      log.error(e);
-
-      return Promise.reject(e);
+function getPublishedFiles(publishedProject) {
+  return PublishedFiles.query({
+    where: {
+      project_id: publishedProject.id
+    }
+  }).fetchAll()
+  .catch(reject);
+}
+
+function clearPublishedProject(publishedProject, mainProject, user) {
+  var publishedFiles;
+
+  // First, get project + files
+  return Promise.resolve()
+  .then(function() {
+    return getPublishedFiles(publishedProject);
+  })
+  .then(function(files) {
+    publishedFiles = files;
+
+    // Then, remove them from s3
+    return modifyPublishedFiles(deleteFile, mainProject, user, files);
+  })
+  .then(function() {
+    // Then, delete them from the DB
+    var deletePublishedProject = publishedProject.destroy();
+    var deletePublishedFiles = Promise.map(publishedFiles, function(file) {
+      return file.destroy();
     });
-};
 
-exports.unpublish = function(project) {
+    return Promise.join(deletePublishedFiles, deletePublishedProject, function() {
+      // We deleted things!
+    });
+  });
+}
+
+exports.publish = function publish(project) {
+  var user;
+
   return getUser(project)
-    .then(function(user) {
-      return modifyFiles(deleteFile, project, user);
-    })
-    .then(function() {
-      return project.set({ publish_url: null }).save();
-    })
-    .then(function() {
-      log.info(
-        'Successfully unpublished ' +
-        project.get('title')
-      );
-    })
-    .catch(function(e) {
-      log.error(e);
-
-      return Promise.reject(e);
+  .then(function(u) {
+    user = u;
+
+    return getPublishedProject(project)
+    .then(function(oldPublishedProject) {
+      if (oldPublishedProject) {
+        return clearPublishedProject(oldPublishedProject, project, user);
+      }
     });
+  })
+  .then(function() {
+    return PublishedProject.forge({
+      project_id: project.get('id'),
+      title: project.get('title'),
+      tags: project.get('tags'),
+      description: project.get('description')
+    }).save();
+  })
+  .then(function(publishedProject) {
+    return File.query({
+      where: {
+        project_id: project.id
+      }
+    }).fetchAll()
+    .then(function(files) {
+      if (files.length === 0) {
+        throw 'No files to be published!';
+      }
+
+      return files.mapThen(function(file) {
+        return PublishedFiles.forge({
```

Comment:
This will fail when updating published files no? You're **always** creating new ones.